posterior_averaging/shared/fed_pa_schedule.py

```diff
@@ -0,0 +1,820 @@
+        ICLR 2021 (https://arxiv.org/abs/2010.05273)
+"""
+
+import collections
+import functools
+from typing import Callable, Collection, Optional, Union
+
+import attr
+import tensorflow as tf
+import tensorflow_federated as tff
+from utils import tensor_utils
+
+from optimization.shared.fed_avg_schedule import ServerState, server_update
+
+
+# Convenience type aliases.
+ModelBuilder = Callable[[], tff.learning.Model]
+OptimizerBuilder = Callable[[float], tf.keras.optimizers.Optimizer]
+ClientMixedinFn = Callable[..., bool]
+ClientMixedinScheduleFn = Callable[[int], ClientMixedinFn]
+ClientUpdateDeltaFn = Callable[..., Collection[tf.Tensor]]
+ClientWeightFn = Callable[..., float]
+LRScheduleFn = Callable[[Union[int, tf.Tensor]], Union[tf.Tensor, float]]
+
+
+def _initialize_optimizer_vars(model: tff.learning.Model,
+                               optimizer: tf.keras.optimizers.Optimizer):
+  """Ensures variables holding the state of `optimizer` are created."""
+  delta = tf.nest.map_structure(tf.zeros_like, _get_weights(model).trainable)
+  model_weights = _get_weights(model)
+  grads_and_vars = tf.nest.map_structure(lambda x, v: (x, v), delta,
+                                         model_weights.trainable)
+  optimizer.apply_gradients(grads_and_vars, name='server_update')
+  assert optimizer.variables()
+
+
+def _get_weights(model: tff.learning.Model) -> tff.learning.ModelWeights:
+  return tff.learning.ModelWeights.from_model(model)
+
+
+@attr.s(eq=False, order=False, frozen=True)
+class DataPassOutput(object):
+  """Structure for outputs returned by a single data pass.
+
+  Fields:
+  - `loss`: A scalar tensor that represents the average loss during the pass.
+  - `num_examples`: A scalar tensor that represents the number of seen examples.
+  - `model_weights_trainable`: A dictionary of model's trainable variables that
+      contain the weights obtained at the end of the data pass.
+  - `model_weights_trainable_sample`: A dictionary of model's trainable
+      variables that contain a candidate posterior sample.
+  """
+  loss = attr.ib()
+  num_examples = attr.ib()
+  model_weights_trainable = attr.ib()
+  model_weights_trainable_sample = attr.ib()
+
+
+@attr.s(eq=False, order=False, frozen=True)
+class DeltaUpdateOutput(object):
+  """Structure for outputs returned by delta update functions.
+
+  This structure contains the sufficient state of the dymanic program used by
+  PostAvg to update weight deltas given a new approximate posterior sample.
+  Each field described below is a dictionary of the same structure as model's
+  trainable variables.
+
+  Fields:
+  - `num_samples`: The number of approximate posterior samples so far.
+  - `weights_delta`: Updates to model's trainable variables.
+  - `weights_sample_mean`: The mean of the all posterior samples so far.
+      `n`-th sample from the mean of the previous `(n - 1)` samples.
+  - `dp_state`: The state of the dynamic programming (DP) represented by a
```

Comment:
DP might be a bit confusing due to the prevalence of differential privacy. I don't think you necessarily need to change it, just wanted to call it out.

Reply:
Good point! Renaming it to `recursion_state`, which should be more descriptive / less confusing.